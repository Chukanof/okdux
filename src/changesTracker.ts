import React from "react";
import { get, intersection } from "lodash";

let trackedFn;

function buildNestedKeys(trackedNested) {
  const res = trackedNested.reduce((acc, el) => {
    const path = el.split(".");
    path.pop();
    const key = path.join(".");
    acc.add(key);
    return acc;
  }, new Set());
  return res;
}

export function checkKeyUsage(fn) {
  fn.deps = [];
  trackedFn = fn;
  const result = fn();
  const exactTrack = fn.deps;
  fn.deps = [];
  walkThrowKeys(result);
  const trackedKeys = fn.deps;
  const trackNestedKeys = buildNestedKeys(trackedKeys);

  const res = [result, exactTrack, trackNestedKeys];
  trackedFn = null;
  fn.deps = null;
  return res;
}

export function getAllKeys(data, key = null) {
  let keys = [];
  key && keys.push(key);
  if (Array.isArray(data)) {
    return keys;
  }

  if (typeof data === "object") {
    for (let i in data) {
      const res = getAllKeys(data[i], (key ? key + "." : "") + i);
      keys = keys.concat(res);
    }
  }
  return keys;
}

const SKIP_WALK_AFTER = 20;

export function walkThrowKeys(data, key = null) {
  if (typeof data === "object") {
    let counter = 0;
    for (let i in data) {
      if (++counter > SKIP_WALK_AFTER) {
        break;
      }
      let prevLen = trackedFn.deps.length;
      const el = data[i];
      let nextLen = trackedFn.deps.length;
      if (prevLen !== nextLen) {
        continue;
      }
      walkThrowKeys(el);
    }
  }
}

export function wrapKeys(keys, data) {
  for (let keyPath of keys) {
    const path = keyPath.split(".");
    // eslint-disable-next-line
    path.reduce((parent, prop) => {
      const obj = get(data, parent, data) || data;

      const valueProp = obj[prop];
      const pathToProp = [...parent, prop];
      if (typeof obj !== "object") {
        return pathToProp;
      }
      Reflect.defineProperty(obj, prop, {
        configurable: true,
        enumerable: true,
        get() {
          trackedFn && trackedFn.deps.push(pathToProp.join("."));
          return valueProp;
        }
      });
      return pathToProp;
    }, []);
  }
}

export class ChangesTracker {
  private trackedDeps: Set<string>;
  private trackedNestedDeps: Set<string>;

  constructor() {
    this.trackedDeps = new Set<string>();
    this.trackedNestedDeps = new Set<string>();
  }

  static hasNestedChanges(nestedKeysToTrack, changedKeys) {
    let hasChanges = false;
    for (let item of nestedKeysToTrack) {
      const res = changedKeys.find(key => key.startsWith(item));
      if (res) {
        hasChanges = true;
        break;
      }
    }
    return hasChanges;
  }

  get trackedDependencies() {
    return [...this.trackedDeps];
  }
  get nestedTrackedDependencies() {
    return [...this.trackedNestedDeps];
  }
  public compute(fn) {
    const [cmpData, deps, nested] = checkKeyUsage(fn);

    deps.forEach(el => this.trackedDeps.add(el));
    nested.forEach(el => this.trackedNestedDeps.add(el));

    return cmpData;
  }

  public clearObservedKeys() {
    this.trackedDeps.clear();
    this.trackedNestedDeps.clear();
  }

  public hasChanges(changedKeys) {
    if (this.trackedDependencies.length === 0) {
      return true;
    }
    return (
      intersection(this.trackedDependencies, changedKeys).length > 0 ||
      ChangesTracker.hasNestedChanges(this.trackedNestedDeps, changedKeys)
    );
  }
}
