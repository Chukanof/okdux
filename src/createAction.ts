import { buildAction, PACreator, AsyncCreator, EmptyOrPayload, Box } from "typesafe-actions";

const createAction = <T>(name: string) => buildAction(name).payload<T>();

export type PlainAction<T> = PACreator<string, T> & { defaultValue: T };
const build = {
  plain: name => {
    return buildAction(name).payload();
  },
  mutator: <T>(defaultValue: T) => (name: string): PlainAction<T> => {
    const action: any = (data = defaultValue) => ({ type: name, payload: data });
    action.defaultValue = defaultValue;
    action.getType = () => name;
    return action;
  },
  async: (payload = d => d, meta = d => d) => name => {
    return buildAction(name).async();
  }
};

export type Unpack<T> = T extends PlainAction<infer F> ? F : any;

function createActions<T extends { [M in keyof T]: T[M] }>(
  actions: T,
  prefix: string = "@"
): { [M in keyof T]: PlainAction<Unpack<ReturnType<T[M]>>> } {
  //@ts-ignore
  return Object.keys(actions).reduce((acc, el) => {
    acc[el] = actions[el](prefix + "/" + el);
    return acc;
  }, {});
}

function createEffects<T>(
  actions: T,
  prefix: string = "@"
): { [M in keyof T]: AsyncCreator<string, any, any, any> } {
  // @ts-ignore
  return Object.keys(actions).reduce((acc, el) => {
    acc[el] = actions[el](prefix + "/" + el);
    return acc;
  }, {});
}

export { createAction, buildAction, build, createActions, createEffects };
