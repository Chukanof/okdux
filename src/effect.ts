import { compose } from "redux";

const typeKey = Symbol();

function assignProperty(object, name, value) {
  Reflect.set(object, name, value);
}
export function createEnhancedAction(name, actionType = Symbol(), thunkEnhancers) {
  let watcher;

  function create(payload) {
    const res = { type: name, payload: payload, [typeKey]: actionType };
    setTimeout(() => watcher && watcher(res), 0);
    return res;
  }

  function getType() {
    return actionType;
  }

  let { mappers, enhancers } = thunkEnhancers || {};

  const enhanced = () => {
    const mapper = compose(...mappers);
    const enhancer = compose(...enhancers);

    const dispatcher = enhancer(cb => cb());

    return args => (dispatch, getState) => {
      dispatcher(() => {
        const action = create(args);
        const payload = mapper(action.payload, getState(), args);
        const res = dispatch({ ...action, payload });
        return res;
      });
    };
  };

  function map(fn) {
    return createEnhancedAction(name, Symbol(), {
      mappers: mappers ? [fn, ...mappers] : [fn],
      enhancers
    });
  }

  function mapFn(fn) {
    return createEnhancedAction(name, Symbol(), {
      mappers: mappers,
      enhancers: enhancers ? [fn, ...enhancers] : [fn]
    });
  }

  function watch(fn) {
    watcher = fn;
  }

  const resultAC = thunkEnhancers ? enhanced() : create;
  assignProperty(resultAC, "create", create);
  assignProperty(resultAC, "watch", watch);
  assignProperty(resultAC, "map", map);
  assignProperty(resultAC, "mapFn", mapFn);
  assignProperty(resultAC, "getType", getType);

  return resultAC;
}

export type Action<T> = {
  type: string;
  payload: T;
};

export type ActionFactory<T> = {
  (args: T): Action<T> | undefined;
  watch(fn: (data: Action<T>) => void): void;
  map<M>(fn: ((data: T) => M)): ActionFactory<M>;
  mapFn(fn: () => void): ActionFactory<T>;
};

export function createAction<T>(name: string): ActionFactory<T>;
export function createAction(name: string) {
  return createEnhancedAction(name);
}
