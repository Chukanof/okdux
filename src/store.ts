import React from "react";
import { get, intersection, uniq, flatten, last } from "lodash";
import { shallowEquals } from "./shallowEquals";
import { ChangesTracker, getAllKeys, wrapKeys } from "./changesTracker";
import { LensCreator, makeLens } from "./lens";

export interface IStore<T> {
  map: <P>(fn: (data: T, ctx: any) => P) => IStore<P>;
}
const identity = d => d;

const dispatchCtx = Symbol("dispatchCtx");

const TYPES = {
  SINGLE_SHALLOW: 1,
  SINGLE_TRACK: 2,
  MULTI_TRACK: 3
};

function mergeKeys(data, store) {
  for (let action in store.handlers) {
    const actionInfo = store.handlers[action];
    const existingPaths = data[action];
    data[action] = existingPaths
      ? [...existingPaths, store.getPath().join(".")]
      : [store.getPath().join(".")];
    if (actionInfo && actionInfo.lens) {
      data[action].push(action => {
        return [...store.getPath(), ...actionInfo.lens(action, makeLens()).path].join(".");
      });
    }
  }
}
function forEachStore(stores, fn) {
  for (let item in stores) {
    if (stores[item]) {
      fn(stores[item]);
      if (stores[item].stores) {
        fn(stores[item].stores);
      }
    }
  }
}

function forEachAction(store, fn) {
  for (let item in store.handlers) {
    fn(store.handlers[item]);
  }
}

export class Store<T> implements IStore<T> {
  reactors = [];
  observers = [];
  keys = {};
  selector;
  currentState;
  root = false;
  initialized = false;
  computed = false;
  type = TYPES.SINGLE_SHALLOW;
  changesTracker: ChangesTracker;

  getState() {
    return this.currentState;
  }

  subscribe(fn) {
    this.reactors.push(fn);
    return () => this.reactors.filter(el => !fn);
  }

  constructor(data, type = TYPES.SINGLE_SHALLOW) {
    this.type = type;
    // @ts-ignore
    this.compose = compose.bind(null, this);
    this.selector = data || identity;
    if (type === TYPES.SINGLE_TRACK) {
      this.changesTracker = new ChangesTracker();
    }
  }

  use(dataOrFn) {
    // @ts-ignore
    const origReducer = this.reducer;
    // @ts-ignore
    this.reducer = (state, action) => {
      const res = origReducer(state, action);
      // @ts-ignore
      this.changedAction = action;
      return res;
    };

    if (typeof dataOrFn === "function") {
      return dataOrFn(this);
    }
    const { subscribe, getState, dispatch } = dataOrFn;
    this.root = true;
    this.initialized = true;
    // this[ctxSymbol].context = context;
    //
    mergeKeys(this.keys, this);
    forEachAction(this, data => {
      data.action._dispatchers.add(dispatch);
    });
    // @ts-ignore
    forEachStore(this.stores, el => {
      // el[ctxSymbol] = this[ctxSymbol];
      mergeKeys(this.keys, el);

      forEachAction(el, data => {
        data.action._dispatchers.add(dispatch);
      });
      this.addStore(el);
      el.initialized = true;
    });
    // const getKeys = this[ctxSymbol].changesMonitor.getChangedKeys;

    const getKeys = (keys, action) => {
      if (!action || !action.type || !keys[action.type]) {
        return [];
      }
      return keys[action.type]
        .map(el => {
          if (typeof el === "function") {
            const res = el(action.payload);

            return res;
          }
          return el;
        })
        .filter(el => !!el);
    };
    subscribe(() => {
      // @ts-ignore
      this.set(getState(), getKeys(this.keys, this.changedAction));
    });
    dispatch({ type: "init" });
    return dataOrFn;
  }
  addStore(store) {
    this.observers.push(store);
  }

  // @ts-ignore
  map(fn, shouldWatchNested) {
    const type = shouldWatchNested ? TYPES.SINGLE_TRACK : TYPES.SINGLE_SHALLOW;
    const store = new Store(fn, type);
    this.addStore(store);
    return store;
  }

  handleChanged(computedData, keys) {
    this.computed = true;
    this.currentState = computedData;
    this.observers.forEach(el => el.run(computedData, keys));

    this.reactors.forEach(fn => fn(computedData));
  }

  run(data, keys) {
    let computedData;
    switch (this.type) {
      case TYPES.SINGLE_TRACK:
        computedData = this.changesTracker.compute(() => this.selector(data));
        if (!this.changesTracker.hasChanges(keys)) {
          return;
        }
        break;
      default:
        computedData = this.selector(data, null);
        break;
    }
    if (shallowEquals(this.getState(), computedData)) {
      return;
    }
    this.handleChanged(computedData, keys);
  }

  set(data, keys) {
    if (this.root) {
      keys = this.getState() ? keys : getAllKeys(data);
      wrapKeys(keys, data);
    }
    this.run(data, keys);
  }
}

export function compose(...stores: IStore<any>[]) {
  const fn: any = stores.pop();
  const store = new Store(fn, TYPES.SINGLE_TRACK);
  function reactor() {
    // @ts-ignore
    if (stores.find(el => !el.computed)) {
      return;
    }
    // @ts-ignore
    store.set(stores.map(el => el.getState()), []);
  }
  stores.forEach(el => {
    // @ts-ignore
    el.subscribe(reactor);
  });
  return store;
}
