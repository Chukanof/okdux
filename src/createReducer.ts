import { get } from "lodash";
import { combineReducers, compose } from "redux";
import im from "object-path-immutable";
import { StandardAction, StandardActionPayload } from "./createAction";
import { LensCreator, makeLens } from "./lens";

// function isReducerBuilder(builder) {
//   return builder && typeof builder === "object" && Reflect.has(builder, reducerPathSymbol);
// }

let identity = <T>(d: T, ..._: any[]): T => d;

interface IReducerBuilder<T> {
  select<RootState>(rootState: RootState): T;
  mapState<R, P>(fn: (state: T, props: P) => R): (root: any, props) => R;
  on<E>(event: StandardAction<E>, handler: (state: T, payload: E) => T): IReducerBuilder<T>;
  on<E, R>(
    event: StandardAction<E>,
    lens: (p: E, prop: LensCreator<T, E>) => LensCreator<R, E>,
    handler: (state: R, payload: E) => R
  ): IReducerBuilder<T>;
}

type Unpacked<T> = T extends IReducerBuilder<infer U>
  ? U
  : T extends StandardAction<infer P> ? P : T;

type R<T> = { [P in keyof T]: Unpacked<T[P]> };

export class BaseReducerBuilder<T> implements IReducerBuilder<T> {
  public handlers = {};
  parent;
  path;
  constructor(public initialState: T) {
    if (typeof initialState === "undefined") {
      throw new Error("initial state should not be undefined");
    }
    this.reducer = this.reducer.bind(this);
  }

  setPath(path) {
    this.path = path;
  }

  getPath() {
    if (this.parent) {
      return this.parent.getPath().concat(this.path);
    } else {
      return this.path ? [this.path] : [];
    }
  }

  // @ts-ignore
  on<E>(action: StandardAction<E>, handlerOrLens, handler = null) {
    if (handler) {
      this.lens(action, handlerOrLens, handler);
      return this;
    } else {
      handler = handlerOrLens;
    }
    if (action === undefined || action === null || !action.getType) {
      throw new Error("action should be an action, got " + action);
    }
    this.handlers[action.getType()] = {
      handler,
      action
    };
    return this;
  }
  lens(action, lens, handler) {
    const propLens = makeLens();
    this.handlers[action.getType()] = {
      handler,
      lens,
      action
    };
  }

  select = <R>(rs: R) => {
    const path = this.getPath();
    return path.length ? get(rs, this.getPath()) : rs;
  };
  // @ts-ignore
  mapState = (fn = identity) => {
    return (state, props) => fn(this.select(state), props, state);
  };

  public reducer = <P>(state: T = this.initialState, action: StandardActionPayload<P>): T => {
    if (!action) {
      return state;
    }
    const { type, payload } = action;
    const handlerObj = this.handlers[type];
    if (handlerObj && handlerObj.handler) {
      if (handlerObj.lens) {
        const path = handlerObj.lens(payload, makeLens()).path;
        const data = get(state, path);

        if (data) {
          const subres = handlerObj.handler(data, payload);
          state = im.set(state, path, subres);
        }
      } else {
        const handler = this.handlers[type].handler;
        state = handler(state, payload, action);
      }
    }

    return state;
  };
}

export type ReducerOrAction = BaseReducerBuilder<any> | StandardAction<any>;
export class CombinedReducer<T extends { [i: string]: ReducerOrAction }> extends BaseReducerBuilder<
  R<T>
> {
  constructor(storesToParse: T) {
    super({} as any);

    const parent = { getPath: this.getPath.bind(this) };
    const stores = {};
    // @ts-ignore
    this.stores = stores;
    Object.keys(storesToParse).forEach(el => {
      let reducer = storesToParse[el];

      // @ts-ignore
      if (reducer && reducer.getType) {
        // @ts-ignore
        reducer = new BaseReducerBuilder(reducer.defaultValue).on(reducer, (_, p) => p);
      }
      stores[el] = reducer;

      // @ts-ignore
      reducer.setPath(el);
      // @ts-ignore
      reducer.parent = parent;
    });

    const reducersMap = Object.keys(stores).reduce((acc, el) => {
      // @ts-ignore
      acc[el] = stores[el].reducer;
      return acc;
    }, {});

    // @ts-ignore
    const nestedReducer = combineReducers(reducersMap);
    const plainReducer = this.reducer;
    // @ts-ignore
    this.reducer = (state = this.initialState, action) => {
      // @ts-ignore
      return plainReducer(nestedReducer(state, action), action);
    };
  }
}

export function createState<T>(data: T) {
  return new BaseReducerBuilder<T>(data);
}
export function combineState<T extends { [i: string]: ReducerOrAction }>(data: T) {
  return new CombinedReducer<T>(data);
}

export { R, Unpacked, IReducerBuilder };
