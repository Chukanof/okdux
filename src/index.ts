import { IReducerBuilder, createState as state, combineState, R } from "./createReducer";
import { createAction, StandardAction } from "./createAction";

export function createState<T>(initialState: T): IReducerBuilder<R<T>> {
  if (initialState === undefined) {
    throw new Error("initial state cannot be undefined");
  }
  let reducer;
  if (typeof initialState === "object") {
    const firstKey = Object.keys(initialState)[0];
    if (
      initialState[firstKey] &&
      (initialState[firstKey].reducer || initialState[firstKey].getType)
    ) {
      // @ts-ignore
      reducer = combineState(initialState);
    } else {
      reducer = state(initialState);
    }
  } else {
    reducer = state(initialState);
  }

  reducer.use = dispatch => use(reducer, dispatch);
  return reducer;
}

function forEachStore(stores, fn) {
  for (let item in stores) {
    if (stores[item]) {
      fn(stores[item]);
      if (stores[item].stores) {
        forEachStore(stores[item].stores, fn);
      }
    }
  }
}

function forEachAction(store, fn) {
  for (let item in store.handlers) {
    fn(store.handlers[item]);
  }
}

function use(store, dispatch) {
  const setDispatch = data => {
    data.action.setDispatch(dispatch);
  };
  forEachAction(store, setDispatch);
  forEachStore(store.stores, el => {
    forEachAction(el, setDispatch);
  });
}

export * from "./createAction";
