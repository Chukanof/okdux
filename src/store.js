import React from "react";
import { get } from "lodash";
import { getKeys, reducerPathSymbol } from "./createReducer";

function shallowEq(a, b) {
  if (Object.is(a, b)) {
    return true;
  }
  if (typeof a !== "object" || b === null || typeof a !== "object" || b === null) {
    return false;
  }
  if (Object.getPrototypeOf(a) !== Object.getPrototypeOf(b)) {
    return false;
  }
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  if (keysA.length !== keysB.length) {
    return false;
  }
  for (let i = 0; i < keysA.length; i++) {
    if (
      !Object.prototype.hasOwnProperty.call(b, keysA[i]) ||
      !Object.is(a[keysA[i]], b[keysA[i]])
    ) {
      return false;
    }
  }
  return true;
}

let trackedFn;
export function checkKeyUsage(data, fn) {
  fn.deps = [];
  trackedFn = fn;
  const result = fn(data);
  trackedFn = null;
  const res = [result, fn.deps];
  fn.deps = null;
  return res;
}

export function wrapKeys(keys, data) {
  for (let keyPath of keys) {
    const path = keyPath.split(".");
    // eslint-disable-next-line
    path.reduce((parent, prop) => {
      const obj = get(data, parent, data) || data;
      const valueProp = obj[prop];
      const pathToProp = [...parent, prop];
      Reflect.defineProperty(obj, prop, {
        configurable: true,
        enumerable: true,
        get() {
          trackedFn && trackedFn.deps.push(pathToProp.join("."));
          return valueProp;
        }
      });
      return pathToProp;
    }, []);
  }
}

const identity = d => d;
export class Store {
  reactors = new Set();
  observers = [];
  selector;
  currentValue;

  constructor(fn = identity) {
    this.selector = fn;
  }

  use = ({ subscribe, getState }) => {
    subscribe(() => {
      this.set(getState(), getKeys());
    });
  };

  react = fn => {
    this.reactors.add(fn);
    return () => this.reactors.filters(el => !fn);
  };

  map = fn => {
    const store = new Store(fn);
    this.observers.push(store);
    return store;
  };
  set = (data, keys) => {
    if (this[reducerPathSymbol]) {
      wrapKeys(keys, data);
    }
    let [computedData, deps] = checkKeyUsage(data, this.selector);
    const hasDeps = deps.length > 0;
    if (hasDeps || !shallowEq(this.value, computedData)) {
      this.value = computedData;
      this.reactors.forEach(fn => fn(computedData));
      this.observers.forEach(el => {
        el.set(computedData, keys);
      });
    }
  };

  getValue() {
    return this.currentValue;
  }

  callReactors(data) {
    const computedData = this.selector(data);
    this.reactors.forEach(fn => fn(computedData));
  }

  tree() {
    return {
      value: this.currentValue,
      children: this.observers.map(el => el.tree())
    };
  }
}

function compose(...stores) {
  const store = new Store();
  function reactor() {
    store.callReactors(stores.map(el => el.getValue()));
  }
  stores.forEach(el => {
    el.react(reactor);
  });

  return store;
}

// const policiesStore = new Store(
//   { policies: { ui: { searchText: "qewr", counter: 0 } }, machines: [] },
//   data => data.policies
// );

// const searchText = policiesStore
//   .map(policies => policies.ui)
//   .map(el => el.searchText);
// const counter = policiesStore
//   .map(policies => policies.ui)
//   .map(el => el.counter);

// const composed = compose(searchText, counter);

// composed.react(data => {
//   console.log("composed", data);
// });
// counter.react(el => console.log("counter", el));

// searchText.react(data => {
//   console.log(data);
// });

// policiesStore.set(
//   { policies: { ui: { searchText: "qewr" } }, machines: [] },
//   new Set(["policies.ui.searchText"])
// );

// let i = 0;
// setInterval(() => {
//   policiesStore.set(
//     { policies: { ui: { searchText: "qewr", counter: ++i } }, machines: [] },
//     new Set(["policies.ui.counter"])
//   );
// }, 6000);

export function createConsumer(store) {
  return class Consumer extends React.Component {
    state = {};
    componentWillMount() {
      this.unsub = store.react(el => {
        this.setState({ state: el });
      });
    }
    componentWillUnount() {
      this.unsub();
    }
    render() {
      return this.props.children(this.state.state);
    }
  };
}

function Consumer({ source }) {
  return createSubscribtion({
    getValue: source.getValue,
    subscribe: source.react
  });
}
export const middleware = (store, getKeys) => ({ dispatch, getState }) => next => action => {
  next(action);
  store.set(getState(), getKeys());
};
// console.log("policiesStore: ", policiesStore.tree());
