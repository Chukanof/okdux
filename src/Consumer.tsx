import * as React from "react";
import { Store } from "./store";

export class Consumer extends React.PureComponent<
  { children: any; source: any; selector: any },
  { currentState: any }
> {
  static displayName = "StoreConsumer";
  state: { currentState: any };

  _unsubscribe: Function;
  _hasUnmounted: boolean = false;
  propStore: any;
  store: any;

  constructor(props) {
    super(props);
    if (props.selector) {
      this.store = props.source.map(state => {
        return props.selector(state, this.props || props);
      }, props.track);
      this.state = { currentState: props.selector(props.source.getState(), props) };
    } else {
      this.store = props.source;
      this.state = { currentState: this.store.getState() };
    }
  }

  componentDidMount() {
    this.subscribe();
  }
  componentWillUnmount() {
    this.unsubscribe();
    this._hasUnmounted = true;
  }
  render() {
    // @ts-ignore
    return this.props.children(this.state.currentState);
  }

  subscribe() {
    const callback = (state: any) => {
      if (this._hasUnmounted) {
        return;
      }

      this.setState({ currentState: state });
    };

    const unsubscribe = this.store.subscribe(callback);

    this._unsubscribe = unsubscribe;
  }
  unsubscribe() {
    if (typeof this._unsubscribe === "function") {
      this._unsubscribe();
    }

    this._unsubscribe = null;
  }
}

export function connect(store, selector) {
  return Component => {
    return props => {
      return (
        <Consumer source={store} selector={selector} {...props}>
          {data => <Component {...props} {...data} />}
        </Consumer>
      );
    };
  };
}
