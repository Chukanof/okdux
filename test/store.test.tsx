import React from "react";
import { createStore } from "redux";
import { Store, compose } from "../src/store";
import { ChangesTracker, checkKeyUsage, wrapKeys, getAllKeys } from "../src/changesTracker";
import { createState, createAction, Consumer, local, StandardAction } from "../src";
import Render from "react-test-renderer";
let mockedData;
let changedKeys;
beforeAll(() => {
  mockedData = {
    ui: { a: { b: 5 } },
    users: { password: { hash: "", data: "" }, name: "" },
    ke: { a: 5 },
    d: 6
  };

  changedKeys = ["ui.a", "users"];
});
describe("restate", () => {
  it("getAllKeys works correctly", () => {
    const keys = getAllKeys(mockedData);
    expect(keys).toMatchSnapshot("getAllKeys");
  });
  it("changesTracker checks correct deps and track changes correctly", () => {
    wrapKeys(getAllKeys(mockedData), mockedData);
    const changesTracker = new ChangesTracker();

    changesTracker.compute(() => {
      return [mockedData.ui.a, mockedData.users.name];
    });

    changesTracker.compute(() => {
      let users = mockedData.users;
      return [mockedData.ui.a, users, mockedData.users.name];
    });

    expect(changesTracker.trackedDependencies).toEqual(["ui", "ui.a", "users", "users.name"]);
    expect(changesTracker.nestedTrackedDependencies).toEqual(["ui.a", "users"]);

    expect(changesTracker.hasChanges(["ui"])).toBeTruthy();
    expect(changesTracker.hasChanges(["ui.a.qwer"])).toBeTruthy();
    expect(changesTracker.hasChanges(["users"])).toBeTruthy();
    expect(changesTracker.hasChanges(["users.kek"])).toBeTruthy();
    expect(changesTracker.hasChanges(["otherkey"])).toBeFalsy();
    expect(changesTracker.hasChanges(["ui.data"])).toBeFalsy();
    changesTracker.clearObservedKeys();
    expect(changesTracker.trackedDependencies).toEqual([]);
    expect(changesTracker.nestedTrackedDependencies).toEqual([]);
  });
  it("calculates computed", () => {
    const inc = createAction("inc");
    const set = createAction("text");

    const counter = createState(1);
    counter.on(inc, state => state + 1);
    const text = createState("hello");
    text.on(set, (data, ev) => data + ev);
    const root = createState({ text, counter });
    const computedValue = counter.map(data => Math.random() >= 0.5);
    const computedValue2 = computedValue.compose(text, data => {
      return { a: [data[0]] };
    });
    const store = root.use(local);
    store.dispatch(inc());

    computedValue.subscribe(el => {});
    computedValue2.subscribe(el => {});

    for (let i = 0; i < 8; i++) {
      store.dispatch(inc());
    }
  });

  it("works ok with store", () => {
    const store = new Store();
    const newData = { ui: { a: "5" }, users: { password: "5dsf", name: "qwer" } };
    const cb = jest.fn();
    const computedStore = store.map(data => ({ ui: data.ui.a, password: data.users.password }));
    computedStore.subscribe(cb);

    store.set(newData, ["ui.a", "users.password"]);
    expect(cb.mock.calls[0][0]).toEqual({ ui: "5", password: "5dsf" });
    store.set({ ...newData, ui: { a: "6" } }, ["users.password"]);
    store.set({ ...newData, ui: { a: "7" } }, ["ui.a"]);

    expect(cb.mock.calls.length).toEqual(3);
  });

  it("works ok with reducer", () => {
    const toggleEvent = createAction("toggle");
    const toggle = createState(0);
    const toggle2 = createState("a");

    toggle.on(toggleEvent, (state, payload) => state + 1);

    const rootState = createState({ toggle, toggle2 });

    const reducer = rootState.reducer;
    const store = createStore(reducer);

    rootState.use(store);
    // expect(rootState.observers.length).toBe(2);

    store.dispatch({ type: "" });
    toggleEvent();
    const renderer = Render.create(
      <Consumer source={toggle}>
        {data => {
          expect(data).toEqual(toggle.getState());
          return <div>{JSON.stringify(data)}</div>;
        }}
      </Consumer>
    );

    const renderer2 = Render.create(
      <Consumer
        source={toggle}
        selector={data => {
          return data + 1;
        }}
      >
        {data => {
          expect(data).toEqual(2);
          return <div>{JSON.stringify(data)}</div>;
        }}
      </Consumer>
    );
  });

  it("rerender only needed times", () => {
    const toggleEvent = createAction("toggle");
    const counterEvent = createAction("counter");
    const toggle = createState(true);
    const counter = createState(0);
    toggle.on(toggleEvent, (state, payload) => !state);
    counter.on(counterEvent, state => state + 1);
    const rootState = createState({ toggle, counter });
    const reducer = rootState.reducer;
    const store = createStore(reducer);

    const toggleViewState = rootState
      .map(state => {
        return {
          toggle: { a: { c: state.toggle } }
        };
      })
      .map(
        el => (
          {
            state: el,
            toggleEvent: e => dispatch(toggleEvent(e))
          },
          true
        )
      );
    rootState.use(store);

    const fn1 = jest.fn();
    const fn2 = jest.fn();
    const fn3 = jest.fn();
    rootState.map(fn3);

    toggleViewState.subscribe(fn1);
    counter.map(el => el).subscribe(fn2);

    store.dispatch({ type: "init" });

    // store.dispatch(toggleEvent());
    // store.dispatch(toggleEvent());
    toggleEvent();
    counterEvent();
    counterEvent();
    counterEvent();
    expect(fn1.mock.calls.length).toBe(1);
    expect(fn2.mock.calls.length).toBe(4);
    expect(fn2.mock.calls).toMatchSnapshot();
    expect(fn1.mock.calls).toMatchSnapshot();
  });

  it("works ok with ministore", () => {
    const e = createAction("e");
    const state = createState("a");
    state.on(e, state => state + 1);
    const fn = jest.fn();

    state.map(fn);

    const e2 = createAction("a");
    const state2 = createState("b");

    state2.on(e2, (_, p) => p);
    const common = createState({ state, state2 });

    state2.subscribe(data => {});
    state.subscribe(() => {});
    const st = common.use(local);

    st.dispatch(e());
    st.dispatch(e2(1));
    st.dispatch(e2(5));

    expect(fn.mock.calls[0][0]).toEqual("a1");
  });

  it("works with lenses", () => {
    const toggle: StandardAction<number> = createAction("TOGGLE");
    const state = createState({ data: [{ id: 1 }, { id: 2, title: "qwer" }], kek: " 5" });
    state.on(
      toggle,
      (payload, prop) => prop.key("data").index(payload),
      (state, p) => {
        return { ...state, title: "qqq" };
      }
    );

    const store = state.use(local);
    const computed = state.map(el => ({ very: { nested: { object: el.data[1] } } }), true);
    const fn = jest.fn();
    computed.subscribe(fn);
    toggle(1);
    toggle(2);
    toggle(2);
    toggle(3);
    toggle(8);

    const reducer = state.reducer;
    expect(reducer(undefined, toggle.raw(1))).toEqual({
      data: [{ id: 1 }, { id: 2, title: "qqq" }],
      kek: " 5"
    });
    expect(fn.mock.calls.length).toBe(1);
  });

  it("works with action binding", () => {
    const toggle: StandardAction<number> = createAction("TOGGLE");
    const state = createState(0);
    state.on(toggle, state => {
      return state + 1;
    });
    const store = state.use(local);
    toggle(1);
    expect([...toggle._dispatchers].length).toBe(1);
  });

  it("composes state with actions corrctly", () => {
    const toggle: StandardAction<number> = createAction("TOGGLE");
    const state = createState(0);
    let composed = createState({
      toggle,
      state
    });
    expect(composed.reducer()).toEqual({ toggle: null, state: 0 });
    composed = createState({
      state,
      toggle
    });
    composed = createState({
      toggle
    });
    expect(composed.reducer()).toEqual({ toggle: null });

    composed = createState({
      data: 2
    });
    expect(composed.reducer()).toEqual({ data: 2 });
  });
});
