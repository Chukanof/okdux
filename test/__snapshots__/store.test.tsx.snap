// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restate getAllKeys works correctly: getAllKeys 1`] = `
Array [
  "ui",
  "ui.a",
  "ui.a.b",
  "users",
  "users.password",
  "users.password.hash",
  "users.password.data",
  "users.name",
  "ke",
  "ke.a",
  "d",
]
`;

exports[`restate rerender only needed times 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`restate rerender only needed times 2`] = `
Array [
  Array [
    true,
  ],
]
`;
